### DISCUSSION TITLE
GroupBox2 - Beta Testing

### DISCUSSION THREAD

**Discussion by elParaguayo** [elParaguayo/qtile-extras#312] [Q&A]
Current PR: #311

If you're interested in testing the widget, please install the `groupbox2` branch of this repo and add the `GroupBox2` widget to your bar.

You should read the documentation which provides a lot of detail as to how the widget works. Docs are available here:
https://qtile-extras.readthedocs.io/en/groupbox2/manual/ref/widgets.html#groupbox2

Areas of feedback requested:
* Bugs!
* Do you like the ability to set rules?
* Is the documentation clear enough? If not, what's missing?
* Line only accepts a single colour. Should we allow multiple colours?
* Line is currently on edge of box. Should its position be adjustable? If so, does it need a different parameter to the margin that is currently used to position the block?
* Image is currently adjusted by margin (similar to block). Should this be changed?
* All other feedback welcome too!

Couple of examples of "advanced" use of this widget showing what's possible:
![gbox](https://github.com/elParaguayo/qtile-extras/assets/946265/3def17f7-e54d-4adc-80e8-cc1f801a8bec)
![gboxpl](https://github.com/elParaguayo/qtile-extras/assets/946265/98d73edc-433c-4bae-93e8-3ffad815fab7)

**Comment by holykremowka2137** [comment]
![Screenshot_2024-01-01_19-56-40](https://github.com/elParaguayo/qtile-extras/assets/152803949/443769a1-1ae0-4eb9-a086-28c57ac4076b)

Why does it show this? I have GroupBox2 installed and im even using that rn, but its still throwing me this error

**Reply by elParaguayo** [reply]
You need to:
```python
from qtile_extras.widget.groupbox2 import GroupBoxRule
```

**Reply by holykremowka2137** [reply]
well, that was dumb XD thank u

**Reply by elParaguayo** [reply]
Well, it's not included in any of my examples either...

**Comment by TheHighGroundWins** [comment]


**Reply by elParaguayo** [reply]
The function just goes in your config.

**Reply by TheHighGroundWins** [reply]
I keep getting this error:
```
2024-12-21 13:02:08,433 ERROR libqtile loop.py:_handle_exception():L62 Exception in event loop:
Traceback (most recent call last):
  File "/usr/lib/python3.12/asyncio/events.py", line 88, in _run
    self._context.run(self._callback, *self._args)
  File "/usr/lib/python3.12/site-packages/libqtile/core/manager.py", line 992, in f
    func(*args)
  File "/usr/lib/python3.12/site-packages/libqtile/bar.py", line 621, in _actual_draw
    self._resize(self._length, self.widgets)
  File "/usr/lib/python3.12/site-packages/libqtile/bar.py", line 464, in _resize
    stretchspace = length - sum(i.length for i in widgets if i.length_type != STRETCH)
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/site-packages/libqtile/bar.py", line 464, in <genexpr>
    stretchspace = length - sum(i.length for i in widgets if i.length_type != STRETCH)
                                ^^^^^^^^
  File "/home/jc/.config/qtile/qtile_extras/widget/decorations.py", line 1081, in length_get
    length = int(self.calculate_length())
                 ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jc/.config/qtile/qtile_extras/widget/groupbox2.py", line 959, in calculate_length
    return sum(box.size for box in self.boxes)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/jc/.config/qtile/qtile_extras/widget/groupbox2.py", line 959, in <genexpr>
    return sum(box.size for box in self.boxes)
               ^^^^^^^^
  File "/home/jc/.config/qtile/qtile_extras/widget/groupbox2.py", line 443, in size
    self._prepare()
  File "/home/jc/.config/qtile/qtile_extras/widget/groupbox2.py", line 330, in _prepare
    self._set_formats()
  File "/home/jc/.config/qtile/qtile_extras/widget/groupbox2.py", line 338, in _set_formats
    rules = [rule for rule in self.rules if rule.match(self)]
                                            ^^^^^^^^^^^^^^^^
  File "/home/jc/.config/qtile/qtile_extras/widget/groupbox2.py", line 219, in match
    matched = self._match(box)
              ^^^^^^^^^^^^^^^^
  File "/home/jc/.config/qtile/qtile_extras/widget/groupbox2.py", line 196, in _match
    if not self.screen & ScreenRule.UNSET:
           ~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~
TypeError: unsupported operand type(s) for &: 'ScreenRule' and 'ScreenRule'
```
When trying to use a custom draw function, could this be due to a wrong version of python being used.

**Reply by elParaguayo** [reply]
Please post your custom function and your rules.

**Reply by TheHighGroundWins** [reply]
this is in my config.py
```
def draw_red_square(box):
    w = 10
    h = 10
    x = (box.size - w) // 2
    y = (box.bar.height - h) // 2
    box.drawer.ctx.rectangle(x, y, w, h)
    box.drawer.set_source_rgb("ff0000")
    box.drawer.ctx.fill()

rules=[
GroupBoxRule(custom_draw=draw_red_square).when(occupied=True)
        ]
```
I accidentally deleted my config file and had to remake it from a screenshot. I don't know what I did before, but now nothing seems to be happening. So at least no big errors.

I don't seem to have any rules before so I made a new one.

I was testing the example draw function before making my own.
